components/codec/src/byte.rs

```diff
@@ -0,0 +1,1188 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{Error, Result};
+
+/// Memory-comparable encoding and decoding utility for bytes.
+pub struct MemComparableByteCodec;
+
+const MEMCMP_GROUP_SIZE: usize = 8;
+const MEMCMP_PADDING: u8 = 0;
+
+impl MemComparableByteCodec {
+    /// Calculates the length of the value after encoding.
+    #[inline]
+    pub fn encoded_len(src_len: usize) -> usize {
+        (src_len / MEMCMP_GROUP_SIZE + 1) * (MEMCMP_GROUP_SIZE + 1)
+    }
+
+    /// Encodes all bytes in the `src` into `dest` in ascending memory-comparable format.
+    ///
+    /// Returns the number of bytes encoded.
+    ///
+    /// `dest` must not overlaps `src`, otherwise encoded results will be incorrect.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there is not enough space in `dest` for writing encoded bytes.
+    ///
+    /// You can calculate required space size via `encoded_len`.
+    pub fn encode_all(src: &[u8], dest: &mut [u8]) -> usize {
+        // Refer: https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format
+
+        assert!(dest.len() >= Self::encoded_len(src.len()));
+
+        unsafe {
+            let mut src_ptr = src.as_ptr();
+            let mut dest_ptr = dest.as_mut_ptr();
+            let src_ptr_end = src_ptr.offset(src.len() as isize);
+
+            // There must be 0 or more zero padding groups and 1 non-zero padding groups
+            // in the output.
+            let zero_padding_groups = src.len() / MEMCMP_GROUP_SIZE;
+
+            // Let's first write these zero padding groups.
+            for _ in 0..zero_padding_groups {
+                ::std::ptr::copy_nonoverlapping(src_ptr, dest_ptr, MEMCMP_GROUP_SIZE);
+                src_ptr = src_ptr.offset(MEMCMP_GROUP_SIZE as isize);
+                dest_ptr = dest_ptr.offset(MEMCMP_GROUP_SIZE as isize);
+
+                dest_ptr.write(!0);
+                dest_ptr = dest_ptr.offset(1);
+            }
+
+            // Then, write the last group, which should never be zero padding.
+            let remaining_size = src_ptr_end.offset_from(src_ptr) as usize;
+            let padding_size = MEMCMP_GROUP_SIZE - remaining_size;
+            let padding_marker = !(padding_size as u8);
+            ::std::ptr::copy_nonoverlapping(src_ptr, dest_ptr, remaining_size);
+            ::std::ptr::write_bytes(
+                dest_ptr.offset(remaining_size as isize),
+                MEMCMP_PADDING,
+                padding_size,
+            );
+            dest_ptr = dest_ptr.offset(MEMCMP_GROUP_SIZE as isize);
+            dest_ptr.write_unaligned(padding_marker);
```

Comment:
I'm not really sure whether writing 1 byte to an arbitrary location of memory can be considered "write aligned" since according to [the documentation](https://doc.rust-lang.org/std/ptr/fn.write_unaligned.html) it looks like the write location should be aligned even if writing zero bytes.

Maybe we can write something like `*dest_ptr = padding_marker` so that the compiler will know what's it?

@kennytm @BusyJay

Reply:
@kennytm  Thanks a lot. Your comment makes it very clear!